src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -110,13 +111,43 @@ public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<Diagnost
         /// </summary>
         /// <param name="compilation">The original compilation.</param>
         /// <param name="analyzers">The set of analyzers to include in future analyses.</param>
+        /// <param name="context"><see cref="AnalyzerHostContext"/> analyzers will run under</param>
+        /// <param name="options">Options that are passed to analyzers.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used to abort analysis.</param>
+        public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerHostContext context, AnalyzerOptions options, CancellationToken cancellationToken)
+            : this(compilation, analyzers, context, new CompilationWithAnalyzersOptions(options, onAnalyzerException: null, analyzerExceptionFilter: null, concurrentAnalysis: true, logAnalyzerExecutionTime: true, reportSuppressedDiagnostics: false), cancellationToken)
+        {
+        }
+
+        /// <summary>
+        /// Creates a new compilation by attaching diagnostic analyzers to an existing compilation.
+        /// </summary>
+        /// <param name="compilation">The original compilation.</param>
+        /// <param name="analyzers">The set of analyzers to include in future analyses.</param>
         /// <param name="analysisOptions">Options to configure analyzer execution.</param>
         public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, CompilationWithAnalyzersOptions analysisOptions)
             : this(compilation, analyzers, analysisOptions, cancellationToken: CancellationToken.None)
         {
         }
 
-        private CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken)
+        /// <summary>
+        /// Creates a new compilation by attaching diagnostic analyzers to an existing compilation.
+        /// </summary>
+        /// <param name="compilation">The original compilation.</param>
+        /// <param name="analyzers">The set of analyzers to include in future analyses.</param>
+        /// <param name="context"><see cref="AnalyzerHostContext"/> analyzers will run under</param>
+        /// <param name="analysisOptions">Options to configure analyzer execution.</param>
+        public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerHostContext context, CompilationWithAnalyzersOptions analysisOptions)
+            : this(compilation, analyzers, context, analysisOptions, cancellationToken: CancellationToken.None)
+        {
+        }
+
+        private CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, CompilationWithAnalyzersOptions analysisOptions, CancellationToken cancellationToken) :
```

Comment:
can we avoid this constructor and have the existing ones invoking this explicitly pass in null context?